### PR TITLE
GH-40843: [Java] Cleanup protobuf-maven-plugin usage

### PR DESCRIPTION
**Pull request by laurentgo**
### Rationale for this change

`protobuf-maven-plugin` usage in Arrow codebase does not follow plugins best practices like sharing the same output directory for different execution or not using test goals for generating test classes

### What changes are included in this PR?

* Add protobuf-maven-plugin plugin to top level pom.xml under pluginManagement to define version and common configuration for all modules
* Remove unnecessary executions of test-compile goal when no test protobufs are present
* Remove use of outputDirectory and clearOutputDirectory and let the plugin choose it for each execution (the default output directory is based on the phase (main vs test) and the language/plugin-id)
* Replace use of compile/compile-custom goals with test-compile/test-compile-custom when generating test protobufs

### Are these changes tested?

As those changes are in the build system, they are covered by the build framework and tests run as part of the build

### Are there any user-facing changes?
None
* GitHub Issue: #40843

### CODE REVIEW

File: java/gandiva/pom.xml [hunk 1]
```diff
@@ -132,18 +132,15 @@
...
                 <groupId>org.xolstice.maven.plugins</groupId>
                 <artifactId>protobuf-maven-plugin</artifactId>
-                <version>0.6.1</version>
-                <configuration>
-                    <protocArtifact>com.google.protobuf:protoc:${protobuf.version}:exe:${os.detected.classifier}
-                    </protocArtifact>
-                    <protoSourceRoot>proto</protoSourceRoot>
-                </configuration>
                 <executions>
+                    <id>src</id>
```
Comment (lidavidm) [inline java/gandiva/pom.xml:136]: This file is causing trouble in some of the CI pipelines

```
Error:  Malformed POM /arrow/java/gandiva/pom.xml: Unrecognised tag: 'id' (position: START_TAG seen ...<executions>\n                    <id>... @136:25)  @ /arrow/java/gandiva/pom.xml, line 136, column 25
 @ 
[INFO] 0 goals, 0 executed
Error:  The build could not read 1 project -> [Help 1]
Error:    
Error:    The project org.apache.arrow.gandiva:arrow-gandiva:16.0.0-SNAPSHOT (/arrow/java/gandiva/pom.xml) has 1 error
Error:      Malformed POM /arrow/java/gandiva/pom.xml: Unrecognised tag: 'id' (position: START_TAG seen ...<executions>\n                    <id>... @136:25)  @ /arrow/java/gandiva/pom.xml, line 136, column 25 -> [Help 2]
```

Comment (laurentgo) [reply java/gandiva/pom.xml:136]: Sorry, I forgot to enable the `arrow-jni` profile when building the project. Will address asap

Comment (laurentgo) [reply java/gandiva/pom.xml]: Pushed the fix

**Comment by lidavidm** [comment]
@github-actions crossbow submit *java*

**Comment by vibhatha** [comment]
@github-actions crossbow submit -g java

**Comment by vibhatha** [comment]
@lidavidm seems like the failing CIs not related, could we re-run and verify?

**Review by vibhatha — APPROVED** [review]
Thanks for working on this, LGTM!

**Review by jbonofre — APPROVED** [review]
LGTM. Maybe we should squash the 2 commits at merge.

**Pull request merged**